include/lsst/afw/image/detail/MaskDict.h

```diff
@@ -0,0 +1,185 @@
+/*
+ * Developed for the LSST Data Management System.
+ * This product includes software developed by the LSST Project
+ * (https://www.lsst.org).
+ * See the COPYRIGHT file at the top-level directory of this distribution
+ * for details of code ownership.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#ifndef LSST_AFW_IMAGE_DETAIL_MASKDICT_H
+#define LSST_AFW_IMAGE_DETAIL_MASKDICT_H
+
+#include <memory>
+#include <map>
+#include "lsst/afw/image/Mask.h"
+
+namespace lsst { namespace afw { namespace image { namespace detail {
+
+/*
+ * MaskDict is the internal object that relates Mask's string plane names
+ * to bit IDs.
+ *
+ * The Mask public API only uses MaskPlaneDict, which is just a typedef
+ * to std::map.  A MaskDict holds this MaskPlaneDict, limiting non-const
+ * access to just the add() and erase() methods in order to maintain
+ * a hash of the full dictionary for fast comparison between MaskDicts.
+ *
+ * In order to maximize consistency of mask plane definitions across different
+ * Mask instances, MaskDict also maintains a global list of all active
+ * MaskDict instances.  When a plane is added to Mask, it is typically
+ * automatically added to *all* active Mask instances that do not already use
+ * that bit (via MaskDict::addMaskPlane) as well as a default MaskDict that is
+ * used when constructing future Mask instances.  In contrast, when a plane is
+ * removed, it typically only affects that Mask, and if that Mask is currently
+ * using the default MaskDict, the default is redefined to a copy prior to
+ * the plane's removal (via detachDefault).
```

Comment:
Maybe?  I'm not fond of this behavior either, but I'm worried about how much would break if we did try to change it.  In any case, it's definitely out of scope for this ticket, and not something I have plans to pursue personally.

Reply:
The whole add/remove mask plane functionality (and the awful names) are legacy from before I joined LSST.  I'd be open to a complete redesign -- we use almost none of this complexity.